### ISSUE TITLE
Modify Elemental Systems Like the ElementalStorageUnit to store in Molar Mass Units

### ISSUE THREAD

**Issue by Master-Foo** [Master-Foo/game#46]
As it stands, the ElementalStorageUnit (ESU), ElementalStorageTank (EST), Refinery, Replicator, Minerals, Chemicals, etc. use generic units for representing how much of an Element it contains.

We need to modify these systems to use Molar Mass.

For instance, Figure out how many Mols of Element X are in a cubic Meter of Bauxite and then store that quantity in the EST.

**Comment by jrslagle** [comment]
Hey @Master-Foo, I'm adding mass and density, and I was wondering why you're initializing name outside the init() and elements inside.

 class Boehmite(chemical.Chemical):
    name = "Boehmite"

    def __init__(self):
        chemical.Chemical.__init__(self)
        self.elements = [

**Comment by Master-Foo** [comment]
It's a complicated story, but there is a good reason for it. Ask me in person and I'll explain.

**Comment by jrslagle** [comment]
@Master-Foo, it looks like the pattern is constants which don't call constructor methods go outside as local variables, and ones that do create class instances are set up in the init() constructor. Am I right?

**Comment by jrslagle** [comment]
Anyway, I'm going to bed without a working commit. Spent most of my time just orienting and looking things up.

**Comment by Master-Foo** [comment]
That's mostly right. You can create class instances outside of the scope of __init__(), but the effect has intended or unintended consequences (depending on your level of ignorance).

Usually, the consequences are unintended and can create some problematic bugs if you don't know what the fuck you are doing.

**Comment by jrslagle** [comment]
So since Bauxite is a heterogeneous rock as complicated as concrete, I'm going to substitute magic numbers for the percentages of each component chemical. Later, the actual composition will be determined by @techwulf via procedural fractals as a part of terrain generation.

**Comment by jrslagle** [comment]
Also on topic, @all, does anyone know of a convenient data type for a list of percentage floats adding up to 1? It's essentially the list of weights in a weighted average, but I've been calling it composition here.

**Comment by Master-Foo** [comment]
I'm not sure what you mean by "a list of percentage floats adding up to 1?"

Do you mean something like this?

``` python
array0 = [0.5, 0.25, 0.1, 0.1, 0.5] #sum(array0) = 1 is good
array1 = [0.5, 0.25, 0.1, 0.1] #sum(array1) = 0.95 is bad
```

I don't think something like that exists. However, we know that there is a discrepancy of 0.05 in array1 so you can just add 5% to each item to normalize the array.

```python
array1_normalized = [0.525, 0.2625, 0.105, 0.105] #sum(array1_normalized) = 0.9975 ALMOST 1. work out some floating point kinks and this should work
```

**Comment by jrslagle** [comment]
@Master-Foo, yes that's exactly what I mean. Another place you see it is the % components that make up a pie chart. I'm hoping there's a python library out there for it. Otherwise I'll be making this datatype myself, along with methods like merge, is_valid, remove[#]. This normalized set of components is the data type I want to use for the chemical makeup of a given volume of rock. It makes the chemistry math easy, and we can build the procedural generation to return a Composition.

**Comment by Master-Foo** [comment]
```python
raw = [0.5, 0.25, 0.1, 0.1]

print(sum(raw))
print(raw)

norm = [i/sum(raw) for i in raw]

print(sum(norm))
print(norm)
```

**Comment by jrslagle** [comment]
That's beautiful...

**Comment by Master-Foo** [comment]
Yeah, it is. Now go read the rant I left you on your last pull request.

**Comment by jrslagle** [comment]
So data structure on composition...
When defining a composition, I'm not sure which to use.

an array of touples
'''
composition.Composition([
    (boehmite.Boehmite(),2),
    (diaspore.Diaspore(),1),
    (gibbsite.Gibbsite(),5)
])

(then)
def __init__(*args):
# normalize input
'''
or a dictionary
'''
composition.Composition([
    ("Boehmite",2),
    ("Diaspore",1),
    ("Gibbsite",5)
])

(then)
def __init__(**kwargs):
# normalize input
'''

**Comment by Master-Foo** [comment]
tuples are probably better because having an instantiation will allow us to query boehmite's properties directly.

Is Composition supposed to be the "ore" we pull out of the ground? It's been a week since I fooled around with what I made as far as minerals, so my memory isn't perfect. But is seems like you are just making another mineral and naming it something else.

I'm fine with this, just make sure we don't have competing objects Remove Minerals if they are now deprecated.

**Comment by jrslagle** [comment]
yea, I want to be able to mention a mineral/chemical/element so you can get to its properties, but without writing war-and-peace. I'm thinking about making a dictionary file that translates Fe = elements.Iron() for instance. You'd just have to be okay with title case for ENUM variables.

**Comment by jrslagle** [comment]
@Master-Foo 
Well I'm not sure why you chose bauxite. It's a rock type, not a mineral, but I'm trying to just go with it.

My organization structure for chemistry is

Composition - a data type. Just a normalized array of touples (chemical/element, float). This describes which chemicals are present in which percentages.

Elements - we're just averaging all isotopes.

Chemicals - groups of atoms. Always have a consistent structure, mass, color, hardness, etc

Mixtures - homogeneous mixtures of Chemicals - has a composition but no internal structure.

Materials - have internal structure via procedural template. That template can return a composition for a queried sub-volume. Textures can be generated by querying the composition of each pixel. Ores can be mined by querying the composition of the volume mined. Each member of a material's composition is a pure chemical.

Objects - 3D meshes made of a chemical, mixture, or material. Have a mass and center of mass.

**Comment by Master-Foo** [comment]
I'm pretty much fine with whatever you can make work. Do what you can and I'll probably have some suggestions on making it more efficient. But right now I just want something that is simple and works.

As far as choosing Bauxite, no particular logic there. I built the whole element, chemical, mineral, refinery elemental storage unit, replicator, etc. in a couple of hours to demonstrate the general plan. It's a spherical cow. It doesn't need to be exact right now. just something we can make work now and improve in the future when it becomes a priority.

So, toss out Bauxite, I'm not married to it.

**Comment by Master-Foo** [comment]
IIRC from geology, a rock (Bauxite) is just a composition of minerals. So, I'm with you in calling Bauxite a composition if that is where you are headed. I'll leave the details up to you since you are more interested.

**Comment by jrslagle** [comment]
I think you're confusing composition and composite.

**Comment by Master-Foo** [comment]
I'm no Rockologist or anything, but I'm sure there is a reason they chose the name composite. Regardless, I'm not as concerned with semantics so much as just having something that works. Call it a parrot the compiler doesn't care.

**Comment by jrslagle** [comment]
@Master-Foo I'm explaining myself because this whole lone ranger thing isn't getting me anywhere. I'm clumsy with python and will need some guidance on data structure, etc. So if you understand what I'm aiming for, then you'll know how to break it down into stepping stones, spherical cows, and such.

**Comment by Master-Foo** [comment]
OK, the secret to spherical cows is to completely ignore the realistic cows. Wipe them from your mind. At best, don't ask yourself how you can make this spherical cow into a realistic cow. Ask yourself how you can make this spherical cow 1 step and ONLY 1 step more complicated.

Because, truthfully, we will never have a 100% completely realistic physical representation of chemistry in the game. What we can do is **evolve** into more complex systems that have the illusion of being **reasonably** accurate from the perspective of the player.

As I see it, the most important thing you can do right now, is not try to accurately model chemistry. But to create the illusion that we are (poorly) modeling chemistry with the intention of creating the illusion that we are (fairly) modeling chemistry in the future.

The rule of thumb is, if you can't do the thing you want to do in 1 sitting and be satisfied, you are doing it wrong.

The one big issue that I think is important is figuring out how many Molar Units are in any given Mineral, Composition, etc. And storing them as Molar Units in the Storage Tanks. Anything else can be made 1 step more complicated in the next commit.

**Comment by Master-Foo** [comment]
Maybe this will help. I haven't really described my intentions, so it's probably reasonable that you are confused.

When someone is mining dirt. For now, I think it's best to say, The only unit of dirt someone pulls out of the ground is 1 Cubic Meter. That 1 Cubic Meter of dirt might be Bauxite, or Sand, or whatever. So, we can look at what 1 Cubic Meter of Bauxite is composed of and easily calculate how many Molar Units of any given Element are in that 1 Cubic Meter.

Start there. When it works, make it 1 step more complicated.

**Comment by Master-Foo** [comment]
Oh, BTW.

The Refinery, ESU, Replicator, and Storage Tanks are now represented in the game itself. Soon, we should be able to throw a rock into the replicator and spit out a solar panel.

Use that as motivation if you want to. The spherical cow you are working on has important consequences.

**Comment by jrslagle** [comment]
I'm saving your secret to spherical cows in its own file, because I'm definitely coming back to that one again, lol.

**Comment by jrslagle** [comment]
I totally agree that I should keep tasks to a difficulty of one sitting. This one, like jumping, or animation, turned into a quagmire of prerequisites. My brain does the equivalent of hard drive thrashing when there are too many prerequisites to find a starting point. Bauxite is complex but I've entered the missing chemicals and hard coded a normalized composition now.

I haven't yet adapted your code for normalizing an array of touples, although I just normalized it myself in excel.

If we're adding 1 m^3 of bauxite to the refinery, then might as well add the rock to the refinery and give the rock the properties of 1 m^3 and material of bauxite. So
-give Rock a volume in m^3 and material

-generate rock with inputs (volume, material)
our_rock = newRock(1, rocks.Bauxite())

-add our_rock to the refinery. 


This is actually getting easy once I lay it out in writing.

**Comment by jrslagle** [comment]
@Master-Foo 
So I was tinkering and was wondering if this uses less resources.

When you listed elements, you were going 
aluminum.Aluminum(),  # calling new instances
oxygen.Oxygen()

Hoping to call their properties later? But I found that the () on the end creates an instance of the class with a memory address. Then I found that I can get at the properties of a chemical, just by passing the class around.

'''
raw_composition = [
            # weathered aluminum variants
            (boehmite.Boehmite, 0.154),
            (diaspore.Diaspore, 0.077),
            (gibbsite.Gibbsite, 0.385),
        ]

for member in raw_composition:
    compound = member[0]
    print(compound.mass)
'''

What are the pros and cons to leaving the () off?

**Comment by jrslagle** [comment]
@Master-Foo adapting your list comprehension for normalization to work on components of a touple took a bit, but I'm there now. It's a bit monstrous for one line though - just like that last run-on-sentence...

from
'''
norm = [i/sum(raw) for i in raw]
'''
to
'''
self.composition = [(part[0], part[1] / sum(i[1] for i in list)) for part in list]
'''

**Comment by jrslagle** [comment]
In a round about way I figured out why you were calling instances of chemicals. If you define name and composition in the constructor as self.name = , then it uses the variable of the parent class. If you define a new name outside the constructor, then that masks and overrides the parent class' variables, and you might as well not even have a parent class at that point. using the parent classes' variables keeps things consistent

... putting things back...

**Comment by Master-Foo** [comment]
@jrslagle I was actually thinking about bringing that up. `boehmite.Boehmite` may be the best way to do this in this case. I was kind of sloppy in my implementation, but using `boehmite.Boehmite()` is generally safer.

The only consideration we need to determine is if in the future Bohmite will ever be more complex. As long as it is "Read Only" though out the whole program, But doing this can also have dirty consequences if we ever start writing to it.

My advice, play around with it and tread lightly. It's cumbersome and bug prone and I'm not entirely sure what the effect will be.

**Comment by jrslagle** [comment]
@Master-Foo, we can always move data members around for mutability if we need to. I don't expect Boehmite to ever need to be written to.

**Comment by jrslagle** [comment]
Oh, I think I misread your question about composition. 

Composition is not a mineral or rock, it's a data type. Every chemical and rock type have a composition as a data member. The composition describes what that rock type is composed of in terms of percentages for each chemical. 

A composite is a kind of rock made of smaller stones cemented together. Bauxite is a composite yes, but I'm treating it like a smooth mixture for now (with a static composition).

**Comment by jrslagle** [comment]
Basically I replaced your chemicals[, ] with composition[(,), (,)]

**Comment by jrslagle** [comment]
@Master-Foo, while I'm on this topic it might be good to let you know my plan for composition. I want EVERY in game entity to have a composition. Meaning even your avatar could jump in the refinery and be reduced to elements. Pick up the solar panel and throw it back in or whatever. I want composition to be used widely, so it needs to be written robustly, which I'll probably need you for. 

The compositions of plants and animals would be kept simple - protein (as a ratio of elements), carbs, fat, water, minerals.

**Comment by Master-Foo** [comment]
Funny story, when I put the Refinery in game. The first time I activated it, it sucked my Avatar in and crashed the game.

**Comment by jrslagle** [comment]
@Master-Foo, which inheritance level do you think mass should go in? Entity? Should we do something like 

Base-object/Entity/Tangeable/Inanimate 

And put mass on Tangeable? Since gases would be entities but not have buoyant mass.

**Comment by Master-Foo** [comment]
Tangible is probably fine
